test/e2e/functional/tests/e2e.go

```diff
@@ -3,355 +3,918 @@ package tests
+							serverPodLabel, serverPodLabel, clientPodLabel, retryInterval, timeout)
+					},
+				},
+			},
+		}
+
+		BeforeEach(func() {
+			config = &ingressnodefwv1alpha1.IngressNodeFirewallConfig{}
+			err := infwutils.LoadIngressNodeFirewallConfigFromFile(config, inftestconsts.IngressNodeFirewallConfigCRFile)
+			Expect(err).ShouldNot(HaveOccurred())
+			config.SetNamespace(OperatorNameSpace)
+			config.SetLabels(testArtifactsLabelMap)
+			err = infwutils.EnsureIngressNodeFirewallConfigExists(testclient.Client, config, timeout)
+			Expect(err).ShouldNot(HaveOccurred())
+			//wait for daemonset to be rolled out
+			infDaemonSet := &appsv1.DaemonSet{}
+			err = daemonset.WaitForDaemonSetReady(testclient.Client, infDaemonSet, OperatorNameSpace,
+				"ingress-node-firewall-daemon", retryInterval, timeout)
+			Expect(err).ShouldNot(HaveOccurred())
+		})
+
+		AfterEach(func() {
+			infwutils.DeleteIngressNodeFirewallConfig(testclient.Client, config)
+		})
+
+		for _, entry := range table {
+			entry := entry
+
+			It(entry.it, func() {
+				var testINFs []*ingressnodefwv1alpha1.IngressNodeFirewall
+				var podCleanups []func()
+
+				podNameIP := make(map[string]ips)
+				podNameObj := make(map[string]*corev1.Pod)
+				nextSourceCIDRsOrder := make(map[string]uint32)
+
+				defer func() {
+					for _, podCleanup := range podCleanups {
+						podCleanup()
 					}
+				}()
+
+				for _, podSetupFn := range entry.getTestPods {
+					podObj, cleanup, err := podSetupFn()
+					Expect(err).ShouldNot(HaveOccurred())
+					var podIPs = ips{}
+					podIPs.setV4IP(podObj.Status.PodIPs)
+					podIPs.setV6IP(podObj.Status.PodIPs)
+					podNameIP[podObj.Name] = podIPs
+					podNameObj[podObj.Name] = podObj
+					podCleanups = append(podCleanups, cleanup)
 				}
-			}
-			Eventually(func() error {
-				err := testclient.Client.Create(context.Background(), rules)
-				return err
-			}, infwutils.Timeout, infwutils.Interval).Should(Succeed())
 
-			key := types.NamespacedName{
-				Name: "rules1",
-			}
-			By("checking Ingress node firewall rules resource is created")
-			Eventually(func() error {
-				err := testclient.Client.Get(context.Background(), key, rules)
-				return err
-			}, infwutils.Timeout, infwutils.Interval).Should(Succeed())
+				// confirm initial connectivity conditions for all protocols defined before IngressNodeFirewall policy application
+				for _, reach := range entry.reachables {
+					reach := reach
 
-			By("checking ingress node firewall nodeState resource is created")
-			nodeStateList := &ingressnodefwv1alpha1.IngressNodeFirewallNodeStateList{}
-			Eventually(func() bool {
-				err := testclient.Client.List(context.Background(), nodeStateList)
-				if err != nil {
-					return false
-				}
-				return len(nodeStateList.Items) == len(nodes.Items)
-			}, infwutils.Timeout, infwutils.Interval).Should(BeTrue())
+					for _, protocol := range entry.protocols {
```

Comment:
this block of code is duplicate at L# 885, use function to avoid code duplication